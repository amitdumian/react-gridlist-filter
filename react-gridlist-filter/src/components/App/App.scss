@mixin breakpoint($point) {
  @if $point == large {
      @media (min-width: 1025px) { @content; }
  }
  @if $point == vsmall {
      @media (max-width: 374px) { @content; }
  }
  @if $point == small {
      @media (max-width: 767px) { @content; }
  }
  @if $point == tablet {
      @media (min-width: 768px) and (max-width: 1024px) { @content; }
  }
  @if $point == mediumToLarge {
    @media (min-width: 768px) and (max-width: 1252px) { @content; }
 }
 
}

.spinner {
  height: 50px;
  width: 50px;
  background-image: url("https://cdnjs.cloudflare.com/ajax/libs/galleriffic/2.0.1/css/loader.gif"); 
}

 /*change template column to update the no. of columns to be displayed in a row */
.item-container {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-column-gap: 30px;
  @include breakpoint(small) {
    grid-template-columns: 1fr;
  }
}

.button {
  width: 150px;
  height: 40px;
  text-align: center;
  margin: auto;
  font-size: 18px;
  font-weight: bold;
  @include breakpoint(small) {
    width: 80px;
    height: 25px;
    font-size: 15px;
  }
}

.button-div
{
  padding-top:10px;
  text-align: center;
  @include breakpoint(small) {
      width: 350px;
  }
  @include breakpoint(mediumToLarge) {
      //width: 400px;
      text-align: center;
  }
}

.textbox {
  width: 300px;
  height: 35px;
  margin: 10px;
  @include breakpoint(small) {
    width: 100%;
    height:20px;
  }
  @include breakpoint(mediumToLarge) {
    width: 200px;
    height: 30px;
  }
  @include breakpoint(tablet) {
    width: 300px;
    height: 35px;
  }
}

