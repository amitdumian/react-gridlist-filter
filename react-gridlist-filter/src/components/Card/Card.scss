@mixin breakpoint($point) {
    @if $point == small {
        @media (max-width: 767px) { @content; }
    }
    @if $point == tablet {
        @media (min-width: 768px) and (max-width: 1024px) { @content; }
    }
    @if $point == mediumToLarge {
      @media (min-width: 768px) and (max-width: 1252px) { @content; }
   }
   
  }

.bold{
    font-weight:bold;
    font-size: 15px;
}

.wrapper-div{
    border:solid 1px grey; 
    width:390px;
    margin:20px;
   @include breakpoint(small) {
        width:340px;
   }

   @include breakpoint(mediumToLarge) {
    width:350px;
   }

 }

.card-info
{
    padding: 0px 20px 10px 50px;
}

.img{
   height:580px;
   width: 350px;
   @include breakpoint(small) {
    width:300px;
    height:500px;
   }

    @include breakpoint(mediumToLarge) {
    width:320px;
    height:550px;
    }

}

.img-span
{
    padding:10px;
}