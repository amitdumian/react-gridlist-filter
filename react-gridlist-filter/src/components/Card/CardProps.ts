export default class CardProps {
     cardData: cardObject;
 }

 // defined model and structure for the json response object
 // some of the fields can be marked optional based on final agreement on the api response model
 export interface cardObject{
    name: string;
    rarity: string;
    type: string;
    cost: number;
    set: setObject;
    collectible: boolean;
    text: string;
    attributes: [];
    unique: boolean;
    imageUrl: string;
    id: string;
 }

 export interface setObject{
    id: string;
    name: string;
    _self: string;
 }
 