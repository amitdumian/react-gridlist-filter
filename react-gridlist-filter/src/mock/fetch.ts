/// fetch resolver to be used for mocking the fetch api
export default function() {
    return Promise.resolve({
      json: () =>
        Promise.resolve([
          {
            name: "Raise Dead",
            rarity: "Legendary",
            type: "Action",
            cost: 2,
            set: {
              "id": "cs",
              "name": "Core Set",
              "_self": "https://api.elderscrollslegends.io/v1/sets/cs"
            },
            collectible: false,
            text: "Summon a random creature from each discard pile.",
            attributes: [
              "Endurance"
            ],
            unique: false,
            imageUrl: "https://images.elderscrollslegends.io/cs/raise_dead.png",
            id: "ce7be2e72d6b06a52e50bed01952801ca4ecfade"
          },
          {
            name: "Raise Dead1",
            rarity: "Legendary",
            type: "Action",
            cost: 2,
            set: {
              "id": "cs",
              "name": "Core Set",
              "_self": "https://api.elderscrollslegends.io/v1/sets/cs"
            },
            collectible: false,
            text: "Summon a random creature from each discard pile.",
            attributes: [
              "Endurance"
            ],
            unique: false,
            imageUrl: "https://images.elderscrollslegends.io/cs/raise_dead1.png",
            id: "ce7be2e72d6b06a52e50bed01952801ca4ecfade"
          }
        ])
   
    })
  }