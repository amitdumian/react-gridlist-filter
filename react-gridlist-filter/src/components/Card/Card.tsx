import React from 'react';
import './Card.scss';
import CardpProps from './CardProps';

function Card(CardProps:any) {
  return (
    <div className='wrapper-div'>
            <span className='img-span'>
              <img src ={CardProps.cardData.imageUrl} className='img' alt = {CardProps.cardData.imageUrl}>
              </img>
            </span>  
            <div className='card-info'>
            <div><span className='bold'>Name: </span>{CardProps.cardData.name}</div> 
            <div><span className='bold'>Text: </span>{CardProps.cardData.text}</div>
            <div><span className='bold'>Set Name: </span>{CardProps.cardData.set && CardProps.cardData.set.name}</div> 
            <div><span className='bold'>Type: </span>{CardProps.cardData.type}</div> 
        </div>
    </div>  

   );
}
export default Card;