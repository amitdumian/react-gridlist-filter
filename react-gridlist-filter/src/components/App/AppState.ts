
import {cardObject} from '../Card/CardProps';
export default class AppState {
    cardData: cardObject[];
    loading?: boolean;
    searchText?: string;
}
