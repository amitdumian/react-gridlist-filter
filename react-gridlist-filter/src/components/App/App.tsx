import React from 'react';
import AppState from './AppState';
import Card from '../Card/Card';
import AppProps from './AppProps';
import {cardObject} from '../Card/CardProps';
import './App.scss';

export default class App extends React.Component<AppProps, AppState> {
  apiUrl = 'https://api.elderscrollslegends.io/v1/cards';
  pageSize = 20; /// contains the no. of records to display initially and on scroll
  startIndex = 0; /// index to keep track of starting index of records in the display
  minSearchTextLength: 3; /// for performance set min limit after which the search will trigger 
  textMaxLength: 50;
  constructor(props:AppProps){
  super(props);
  this.state = new AppState();
    this.state = {
      cardData:[],
      searchText: ''
    }
  }

  componentDidMount() {
    this.fetchCards();
    window.addEventListener('scroll', this.handleScroll); /// add window scroll event listener 
   }

   // catch and log any script errors
   componentDidCatch(error, info) {
    console.log('error: ', error);
    console.log('info: ', info);
   }
   
   /// Remove the event listner on component unmount
   componentWillUnmount() {
     window.removeEventListener('scroll', this.handleScroll)
   }

   fetchCardsOnScroll(startIndex:number)
   {
     const fetchPromise = fetch(this.apiUrl);
     this.setState({ loading: true }) // start spinner
     fetchPromise.then(response => response.json())
      .then(data => {
        const offers = data.cards.slice();  // pick up the original cards returned from api
        const filteredCards = this.getFilterCards(offers);
        if(filteredCards.length > this.pageSize)
        {
          const moreData = filteredCards.splice(startIndex, this.pageSize); // fetch  subset of records
          if (moreData.length > 0 && (startIndex+this.pageSize <= offers.length)) {
            this.startIndex =  startIndex;
            this.setState({ cardData: this.state.cardData.concat(moreData) });
          }
        }
         this.setState({ loading: false })
        } 
      )
      .catch((error) => {
        console.log(error);
        this.setState({ loading: false }); /// stop spinner
        return Promise.reject();
      });
   }


  /// method gets call on page load
  /// when text entered in the box
  /// When button is clicked for search
   fetchCards = () =>
   {
     this.setState({ loading: true })
     const fetchPromise = fetch(this.apiUrl);
     fetchPromise
      .then(response => response.json())
      .then(data => {
          let filteredCards = data.cards;
          // if search text is passed, filter the cards based on name
          if(this.state.searchText && this.state.searchText.trim() !== '')
          {
            filteredCards = this.getFilterCards(data.cards);
          }
          this.setState({ cardData: filteredCards.splice(this.startIndex, this.pageSize) })
          this.setState({ loading: false })
        }
      )
      .catch((error) => {
        console.log(error);
        this.setState({ loading: false })
        return Promise.reject()
      });
   
   }

   // function to return the list of filtered cards based on card name
   getFilterCards = (cards: cardObject[]) =>
   {
        // data field based on which records need to be filtered
        const searchText = this.state.searchText?this.state.searchText:'';
        const searchField = 'name';
        const filteredCards = cards.filter(function(product:cardObject) {
              for (var i in product) {
                  if (product[searchField].toLowerCase().indexOf(searchText.toLowerCase()) > -1) 
                  return product;
              }
         })

        return filteredCards;
   }

  
   // function called on textbox onchange event
   handleChange = (event:any) => {
      this.setState({searchText: event.target.value});
      //if(event.target.value && event.target.value.length > this.minSearchTextLength)
      //{
         this.fetchCards();
      //}
    }
  
   // function call triggered on window scroll
   handleScroll = (e:any) => {
    let element = e.target;
    if (element.scrollingElement.scrollHeight - element.scrollingElement.scrollTop === element.scrollingElement.clientHeight) {
      this.fetchCardsOnScroll(this.startIndex+this.pageSize);
     }
   }

  render() {
   const{cardData} = this.state;
      return (
      <React.Fragment>
           <div className='button-div'> 
              <input type='text'
                placeholder="Search by name"  
                value={this.state.searchText}
                maxLength = {this.textMaxLength}
                onChange={this.handleChange} 
                className='textbox'
              />
             <button onClick={this.fetchCards} className='button'>Search</button>
            </div>
             {this.state.loading && this.renderSpinner()}
             {cardData && this.renderGrid()}
      </React.Fragment>
    
    );
  }

  // function to return the spinner div html
  renderSpinner()
  {
    return(
      <div className = 'spinner'></div>
    )
  }


  renderGrid() {
   const cData= this.state.cardData;
   return(
      <div className = 'item-container'>
          {cData.map((data:cardObject, i:number) => (
            <React.Fragment>
              <Card 
                cardData={data}>
              </Card>
            </React.Fragment>
            ))}
      </div>
    )
  }
 }
 


