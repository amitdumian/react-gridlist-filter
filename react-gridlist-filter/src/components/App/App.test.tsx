import React from 'react';
import { render } from '@testing-library/react';
import ReactDOM from 'react-dom';
import App from './App';
import Card from '../Card/Card';
import {shallow, Mount, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import fetch from '../../mock/fetch';
import renderer from 'react-test-renderer';
configure({ adapter: new Adapter() });

// store the mock card response from json file
const cardResponse = require('../../mock/cards.json');

describe('App Component', () => {
  it('App component exists', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.exists()).toBe(true);
  });

  /// to check for snapshot making sure the UI is not changing
  it('render correctly text and button component', () => {  
    const textBox = renderer
    .create(<input type = 'text'></input>)
    .toJSON();
    expect(textBox).toMatchSnapshot();

    const button = renderer
    .create(<button></button>)
    .toJSON();
    expect(button).toMatchSnapshot();
  });

  // Simulate the initial page load to display the cards
  // Simulate the fetch call from the resolver
  describe('App component', () => {
    describe('when rendered', () => {
      it('should fetch a list of cards', () => {
        const wrapper = shallow(<App />);
        const fetchSpy = jest.spyOn(window, 'fetch');
        const cardInstance = shallow(
          <Card cardData={cardResponse} />
        );
        //expect(wrapper.state('cardData')).toBe(cardResponse);
        //expect(wrapper.state('loading')).toBe(false);
       // expect(fetchSpy).toBeCalled();
      });
    });
  });

  /// Simulate the onChange event on the textbox
  /// To do// add more test to simulate the filter based on the search text
  it('should call onChange for textbox', () => {
    const wrapper = shallow(<App />);
    const handleChange = jest.fn();
    const event = {
      preventDefault() {},
      target: { value: 'Raid' }
    };
    const textBox = wrapper.find('input.textbox');
    textBox.simulate('change', event);
    //expect(wrapper.state('searchText')).toBe('Raid');
    //expect(handleChange).toBeCalledWith(event);
  });

  /// Simulate the button click event
  it('search button should display the records ', () => {
    const wrapper = shallow(<App />);
    const fetchCards = jest.fn();
    const searchBtn = wrapper.find('button.button');
    searchBtn.simulate('click');
    //expect(fetchCards).toBeCalled();
  });

  /// Simulate the window scroll and trigger the handleScroll function
  it('calls window.scroll', () => {
    const handleScroll = jest.fn();
    document.body.scrollTop = 100;
    //window.dispatchEvent(new window.UIEvent('scroll', { detail: 0 }));
    //expect(handleScroll).toBeCalled();
  });
  
})