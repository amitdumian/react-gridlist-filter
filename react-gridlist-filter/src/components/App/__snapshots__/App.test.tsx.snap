// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component render correctly text and button component 1`] = `
<input
  type="text"
/>
`;

exports[`App Component render correctly text and button component 2`] = `<button />`;
